modin/core/dataframe/pandas/partitioning/partition_manager.py

```diff
@@ -1329,35 +1329,31 @@ def compute_schema(axis_lengths, axis):
             tuple
                 Tuple of (schema, updated axis lengths aсcording to the schema)
             """
-            from modin.core.storage_formats.pandas.utils import (
-                compute_chunksize,
-            )
-
             idx = pandas.RangeIndex(sum(axis_lengths))
+            # It will be possible to use axis length for `compute_chunksize` instead of dataframe
+            # after merge of https://github.com/modin-project/modin/issues/3768 happens.
             df = pandas.DataFrame(**{"columns" if axis == 1 else "index": idx})
             chunksize = compute_chunksize(df, NPartitions.get(), axis=axis)
 
             schema = []
-            # initialize the first interval
-            schema.append([0, None])
             current_axis_length = 0
 
             for idx, axis_length in enumerate(axis_lengths):
+                if current_axis_length == 0:
+                    # create new interval
+                    schema.append([idx, None])
                 if current_axis_length + axis_length < chunksize:
                     current_axis_length += axis_length
                     continue
                 current_axis_length = 0
                 # update end of the interval
                 schema[-1][1] = idx + 1
-                if idx + 1 != len(axis_lengths):
-                    schema.append([idx + 1, None])
+
             if schema[-1][1] is None:
                 schema[-1][1] = len(axis_lengths)
 
             # update axis lengths aсcording to the schema
-            new_axis_lengths = list(
-                map(lambda t: sum(axis_lengths[t[0] : t[1]]), schema)
-            )
+            new_axis_lengths = [sum(axis_lengths[t[0] : t[1]]) for t in schema]
```

Comment:
I just wonder why you first compute indices and then sum lengths instead of computing the lengths of new partitions along the way - you compute the length anyway above to compute the indices.

Reply:
This solution is shorter (one line shorter) and better readable. It's clear that unnecessary summation is taking place here, but I think it will not take most of the total repartition time :)